sway-core/src/type_system/engine.rs

```diff
@@ -20,6 +20,12 @@ pub struct TypeEngine {
     id_map: RwLock<HashMap<TypeInfo, TypeId>>,
 }
 
+enum UnifyKind {
```

Comment:
This needs extensive doccomments to explain that the differences between the unification modes is.